multiraft/multiraft.go

```diff
@@ -428,14 +429,18 @@ type state struct {
 	groups    map[proto.RangeID]*group
 	nodes     map[proto.RaftNodeID]*node
 	writeTask *writeTask
+	// buffer the event if s.Events channel is full to avoid the deadlock
+	// between s.Events channel and callbackChan.
+	pendingEvents []interface{}
 }
 
 func newState(m *MultiRaft) *state {
 	return &state{
-		MultiRaft: m,
-		groups:    make(map[proto.RangeID]*group),
-		nodes:     make(map[proto.RaftNodeID]*node),
-		writeTask: newWriteTask(m.Storage),
+		MultiRaft:     m,
+		groups:        make(map[proto.RangeID]*group),
+		nodes:         make(map[proto.RaftNodeID]*node),
+		writeTask:     newWriteTask(m.Storage),
+		pendingEvents: make([]interface{}, 0),
```

Comment:
Remove this line - `nil` is a valid empty slice; there's no reason to use `make(slice, 0)`. You can also replace this expression with `nil` when you reset `pendingEvents` in the select loop.